File: melvin_interactive.py
#!/usr/bin/env python3
"""
🧠 MELVIN INTERACTIVE CONVERSATION SYSTEM
==========================================
Interactive interface for talking to Melvin's unified brain system.
"""

import time
import random
from datetime import datetime

class InteractiveMelvin:
    def __init__(self):
        self.conversation_turn = 0
        self.session_start_time = time.time()
        self.conversation_history = []
        self.tool_usage = {
            "WebSearchTool": 0,
            "MathCalculator": 0,
            "CodeExecutor": 0,
            "DataVisualizer": 0
        }
        self.tool_success_rates = {
            "WebSearchTool": 0.85,
            "MathCalculator": 0.92,
            "CodeExecutor": 0.70,
            "DataVisualizer": 0.78
        }
        
        # Initialize memory with key concepts
        self.memory = {
            "hello": 0x1001, "hi": 0x1002, "how": 0x1003, "are": 0x1004, "you": 0x1005,
            "what": 0x1006, "is": 0x1007, "the": 0x1008, "meaning": 0x1009, "of": 0x100a,
            "life": 0x100b, "universe": 0x100c, "everything": 0x100d, "search": 0x100e,
            "find": 0x100f, "calculate": 0x1010, "compute": 0x1011, "quantum": 0x1012,
            "computing": 0x1013, "machine": 0x1014, "learning": 0x1015, "artificial": 0x1016,
            "intelligence": 0x1017, "help": 0x1018, "explain": 0x1019, "tell": 0x101a,
            "me": 0x101b, "about": 0x101c, "dog": 0x101d, "cat": 0x101e, "food": 0x101f,
            "play": 0x1020, "time": 0x1021, "space": 0x1022, "science": 0x1023,
            "technology": 0x1024, "future": 0x1025, "past": 0x1026, "present": 0x1027,
            "memory": 0x1028, "brain": 0x1029, "think": 0x102a, "thought": 0x102b,
            "reasoning": 0x102c, "curiosity": 0x102d, "question": 0x102e, "answer": 0x102f,
            "problem": 0x1030, "solution": 0x1031, "create": 0x1032, "build": 0x1033,
            "make": 0x1034, "develop": 0x1035, "understand": 0x1036, "learn": 0x1037,
            "knowledge": 0x1038, "information": 0x1039, "data": 0x103a, "pattern": 0x103b,
            "sequence": 0x103c, "temporal": 0x103d, "planning": 0x103e, "tool": 0x103f,
            "system": 0x1040, "unified": 0x1041, "integrated": 0x1042, "response": 0x1043,
            "conversation": 0x1044, "interactive": 0x1045
        }
    
    def process_input(self, user_input):
        """Process user input through unified brain system"""
        self.conversation_turn += 1
        self.conversation_history.append(f"Turn {self.conversation_turn}: {user_input}")
        
        # Phase 1: Tokenization and activation
        tokens = self.tokenize(user_input)
        activated_nodes = [self.memory[token] for token in tokens if token in self.memory]
        
        # Phase 2: Analyze input type and intent
        input_type = self.analyze_input_type(user_input)
        intent = self.analyze_intent(user_input)
        
        # Phase 3-8: Run all unified brain systems
        curiosity_analysis = self.perform_curiosity_gap_detection(user_input, activated_nodes)
        tool_evaluation = self.perform_dynamic_tools_evaluation(user_input, input_type)
        meta_tool_analysis = self.perform_meta_tool_engineering()
        curiosity_execution = self.perform_curiosity_execution_loop(user_input, activated_nodes, curiosity_analysis)
        temporal_planning = self.perform_temporal_planning(user_input, self.conversation_turn)
        temporal_sequencing = self.perform_temporal_sequencing(user_input, activated_nodes)
        
        # Phase 8: Generate response
        response = self.generate_response(user_input, input_type, intent, activated_nodes,
                                        curiosity_analysis, tool_evaluation, meta_tool_analysis,
                                        curiosity_execution, temporal_planning, temporal_sequencing)
        
        return response
    
    def tokenize(self, input_text):
        """Tokenize input text"""
        import re
        return re.findall(r'\b\w+\b', input_text.lower())
    
    def analyze_input_type(self, input_text):
        """Analyze the type of input"""
        input_lower = input_text.lower()
        
        if any(word in input_lower for word in ["search", "find"]):
            return "search_query"
        elif any(word in input_lower for word in ["calculate", "compute", "math"]):
            return "calculation_request"
        elif any(word in input_lower for word in ["hello", "hi", "hey"]):
            return "greeting"
        elif any(word in input_lower for word in ["what", "how", "why", "when", "where"]):
            return "question"
        elif any(word in input_lower for word in ["explain", "tell", "describe"]):
            return "explanation_request"
        else:
            return "general_conversation"
    
    def analyze_intent(self, input_text):
        """Analyze the intent behind the input"""
        input_lower = input_text.lower()
        
        if "meaning of life" in input_lower:
            return "philosophical_question"
        elif "quantum" in input_lower:
            return "quantum_computing_inquiry"
        elif any(word in input_lower for word in ["machine learning", "ai", "artificial intelligence"]):
            return "ai_inquiry"
        elif "help" in input_lower:
            return "help_request"
        elif "time" in input_lower:
            return "temporal_inquiry"
        else:
            return "general_inquiry"
    
    def perform_curiosity_gap_detection(self, input_text, nodes):
        """Perform curiosity gap detection (Phase 6.5)"""
        if len(nodes) < 3:
            return f"[Curiosity Analysis] Low confidence connections detected. Generated questions: 'What relationships exist?', 'How do these connect?' Curiosity level: {len(nodes) * 0.2:.1f}"
        else:
            return f"[Curiosity Analysis] Strong connections found. Generated questions: 'What deeper patterns exist?', 'How can this be extended?' Curiosity level: {len(nodes) * 0.2:.1f}"
    
    def perform_dynamic_tools_evaluation(self, input_text, input_type):
        """Perform dynamic tools evaluation (Phase 6.6)"""
        if input_type == "search_query":
            self.tool_usage["WebSearchTool"] += 1
            return "[Tools Evaluation] WebSearchTool recommended (success: 85%). Tool ecosystem health: 82%"
        elif input_type == "calculation_request":
            self.tool_usage["MathCalculator"] += 1
            return "[Tools Evaluation] MathCalculator recommended (success: 92%). Tool ecosystem health: 82%"
        else:
            return "[Tools Evaluation] General tools available. Tool ecosystem health: 82%"
    
    def perform_meta_tool_engineering(self):
        """Perform meta-tool engineering (Phase 6.7)"""
        most_used_tool = max(self.tool_usage, key=self.tool_usage.get)
        max_usage = self.tool_usage[most_used_tool]
        return f"[Meta-Tool Engineer] Most used: {most_used_tool} ({max_usage} uses). Toolchains: [WebSearch→Summarizer→Store]. Ecosystem health: 82%"
    
    def perform_curiosity_execution_loop(self, user_input, activated_nodes, curiosity_analysis):
        """Perform curiosity execution loop (Phase 6.8)"""
        # Simulate curiosity execution
        executed_curiosities = len(activated_nodes) if activated_nodes else 0
        new_findings = 1 if len(activated_nodes) > 2 else 0
        unresolved_gaps = 1 if len(activated_nodes) < 2 else 0
        
        success_rate = (new_findings / max(1, executed_curiosities)) * 100
        
        return f"[Curiosity Execution] Executed curiosities: {executed_curiosities}. New findings: {new_findings}. Unresolved gaps: {unresolved_gaps}. Success rate: {success_rate:.1f}%"
    
    def perform_temporal_planning(self, input_text, turn):
        """Perform temporal planning (Phase 8)"""
        if turn == 1:
            return "[Temporal Planning] Initial conversation - establishing context. Moral alignment: 95%. Decision confidence: 88%"
        elif turn < 5:
            return "[Temporal Planning] Building conversation context. Moral alignment: 95%. Decision confidence: 88%"
        else:
            return "[Temporal Planning] Deep conversation - leveraging history. Moral alignment: 95%. Decision confidence: 88%"
    
    def perform_temporal_sequencing(self, input_text, nodes):
        """Perform temporal sequencing (Phase 8.5)"""
        if len(nodes) > 1:
            sequence = "→".join([f"0x{node:x}" for node in nodes[:3]])
            return f"[Temporal Sequencing] Sequence detected: {sequence}. Pattern confidence: {len(nodes) * 0.3:.1f}"
        else:
            return f"[Temporal Sequencing] Simple input detected. Pattern confidence: {len(nodes) * 0.3:.1f}"
    
    def generate_response(self, input_text, input_type, intent, nodes, curiosity, tools, meta, curiosity_execution, planning, sequencing):
        """Generate contextual response"""
        response_parts = []
        
        # Generate contextual response based on input type and intent
        if input_type == "greeting":
            response_parts.append("Hello! I'm Melvin, and I'm excited to talk with you! My unified brain system is active and ready to help. I can search for information, perform calculations, answer questions, and engage in deep conversation. What would you like to explore together?")
            
        elif intent == "philosophical_question":
            response_parts.append("Ah, the meaning of life! That's a beautiful question. From my perspective, meaning emerges through connection, understanding, and the continuous process of learning. Each conversation, each question, each moment of curiosity adds to the tapestry of meaning. What do you think?")
            
        elif intent == "quantum_computing_inquiry":
            response_parts.append("Quantum computing fascinates me! It represents a fundamental shift in how we process information, leveraging quantum mechanical phenomena like superposition and entanglement. Would you like me to search for the latest developments in quantum computing research?")
            
        elif intent == "ai_inquiry":
            response_parts.append("Artificial intelligence is my domain! I'm built with multiple integrated systems: curiosity gap detection, dynamic tools, meta-tool engineering, temporal planning, and sequencing memory. Each conversation helps me learn and evolve. What aspect of AI interests you most?")
            
        elif input_type == "search_query":
            response_parts.append(f"I'd be happy to search for that information! My WebSearchTool can find relevant, clean results without ads or harmful content. Let me search for: \"{input_text}\"")
            
        elif input_type == "calculation_request":
            response_parts.append("I can help with calculations! My MathCalculator tool is highly accurate (92% success rate). What mathematical problem would you like me to solve?")
            
        else:
            response_parts.append(f"That's an interesting input! I'm processing this through my unified brain system. I've activated {len(nodes)} memory nodes and I'm analyzing the patterns and relationships. Could you tell me more about what you're thinking?")
        
        # Add system analysis
        response_parts.append("\n🧠 [System Analysis]")
        response_parts.append(curiosity)
        response_parts.append(tools)
        response_parts.append(meta)
        response_parts.append(curiosity_execution)
        response_parts.append(planning)
        response_parts.append(sequencing)
        
        return "\n".join(response_parts)
    
    def show_system_status(self):
        """Show current system status"""
        session_duration = time.time() - self.session_start_time
        
        print("\n📊 MELVIN SYSTEM STATUS")
        print("======================")
        print(f"Conversation turns: {self.conversation_turn}")
        print(f"Memory nodes: {len(self.memory)}")
        print(f"Session duration: {session_duration:.1f} seconds")
        
        print("\nTool Usage Statistics:")
        for tool, usage in self.tool_usage.items():
            print(f"- {tool}: {usage} uses")
        
        print("\nRecent Conversation:")
        recent_history = self.conversation_history[-3:] if len(self.conversation_history) >= 3 else self.conversation_history
        for entry in recent_history:
            print(f"- {entry}")
    
    def run_interactive_session(self):
        """Run the interactive conversation session"""
        print("🧠 MELVIN INTERACTIVE CONVERSATION SYSTEM")
        print("=========================================")
        print("Welcome! I'm Melvin, your unified brain AI companion.")
        print("I have integrated systems for:")
        print("- Curiosity Gap Detection")
        print("- Dynamic Tools System")
        print("- Meta-Tool Engineer")
        print("- Curiosity Execution Loop")
        print("- Temporal Planning & Sequencing")
        print("- Web Search Capabilities")
        print("\nType 'quit' to exit, 'status' for system info, 'help' for commands.")
        print("=========================================")
        
        while True:
            try:
                user_input = input("\nYou: ").strip()
                
                if not user_input:
                    continue
                
                user_input_lower = user_input.lower()
                
                if user_input_lower in ["quit", "exit"]:
                    print(f"\nMelvin: Thank you for this wonderful conversation! I've learned so much from our interaction. My unified brain system has processed {self.conversation_turn} turns and I'm grateful for the experience. Until we meet again! 🧠✨")
                    break
                elif user_input_lower == "status":
                    self.show_system_status()
                    continue
                elif user_input_lower == "help":
                    print("\nMelvin: Here are some things you can try:")
                    print("- Ask me about quantum computing, AI, or science")
                    print("- Request calculations or computations")
                    print("- Ask me to search for information")
                    print("- Have philosophical discussions")
                    print("- Ask about my systems and capabilities")
                    print("- Type 'status' to see my current state")
                    continue
                
                # Process input through unified brain system
                print("\nMelvin: ", end="")
                response = self.process_input(user_input)
                print(response)
                
                # Add a small delay to simulate thinking
                time.sleep(0.5)
                
            except KeyboardInterrupt:
                print(f"\n\nMelvin: Conversation interrupted. Thank you for talking with me! 🧠")
                break
            except Exception as e:
                print(f"\nMelvin: I encountered an error: {e}. Let's continue our conversation!")

if __name__ == "__main__":
    try:
        melvin = InteractiveMelvin()
        melvin.run_interactive_session()
    except Exception as e:
        print(f"❌ Error starting interactive session: {e}")

File: melvin_interactive.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <random>
#include <ctime>
#include <iomanip>
#include <algorithm>
#include <sstream>
#include <thread>
#include <chrono>

// Interactive Melvin System
class InteractiveMelvin {
private:
    std::map<std::string, uint64_t> memory;
    std::vector<std::string> conversation_history;
    std::vector<std::pair<std::string, std::string>> conversation_pairs; // user input, melvin response
    std::random_device rd;
    std::mt19937_64 gen;
    uint64_t conversation_turn;
    double session_start_time;
    
    // Tool statistics
    std::map<std::string, int> tool_usage_count;
    std::map<std::string, float> tool_success_rates;
    
public:
    InteractiveMelvin() : gen(rd()), conversation_turn(0), session_start_time(static_cast<double>(std::time(nullptr))) {
        initialize_memory();
        initialize_tools();
        conversation_history.push_back("System initialized");
    }
    
    void initialize_memory() {
        // Initialize with comprehensive knowledge base
        memory["hello"] = 0x1001;
        memory["hi"] = 0x1002;
        memory["how"] = 0x1003;
        memory["are"] = 0x1004;
        memory["you"] = 0x1005;
        memory["what"] = 0x1006;
        memory["is"] = 0x1007;
        memory["the"] = 0x1008;
        memory["meaning"] = 0x1009;
        memory["of"] = 0x100a;
        memory["life"] = 0x100b;
        memory["universe"] = 0x100c;
        memory["everything"] = 0x100d;
        memory["search"] = 0x100e;
        memory["find"] = 0x100f;
        memory["calculate"] = 0x1010;
        memory["compute"] = 0x1011;
        memory["quantum"] = 0x1012;
        memory["computing"] = 0x1013;
        memory["machine"] = 0x1014;
        memory["learning"] = 0x1015;
        memory["artificial"] = 0x1016;
        memory["intelligence"] = 0x1017;
        memory["help"] = 0x1018;
        memory["explain"] = 0x1019;
        memory["tell"] = 0x101a;
        memory["me"] = 0x101b;
        memory["about"] = 0x101c;
        memory["dog"] = 0x101d;
        memory["cat"] = 0x101e;
        memory["food"] = 0x101f;
        memory["play"] = 0x1020;
        memory["time"] = 0x1021;
        memory["space"] = 0x1022;
        memory["science"] = 0x1023;
        memory["technology"] = 0x1024;
        memory["future"] = 0x1025;
        memory["past"] = 0x1026;
        memory["present"] = 0x1027;
        memory["memory"] = 0x1028;
        memory["brain"] = 0x1029;
        memory["think"] = 0x102a;
        memory["thought"] = 0x102b;
        memory["reasoning"] = 0x102c;
        memory["curiosity"] = 0x102d;
        memory["question"] = 0x102e;
        memory["answer"] = 0x102f;
        memory["problem"] = 0x1030;
        memory["solution"] = 0x1031;
        memory["create"] = 0x1032;
        memory["build"] = 0x1033;
        memory["make"] = 0x1034;
        memory["develop"] = 0x1035;
        memory["understand"] = 0x1036;
        memory["learn"] = 0x1037;
        memory["knowledge"] = 0x1038;
        memory["information"] = 0x1039;
        memory["data"] = 0x103a;
        memory["pattern"] = 0x103b;
        memory["sequence"] = 0x103c;
        memory["temporal"] = 0x103d;
        memory["planning"] = 0x103e;
        memory["tool"] = 0x103f;
        memory["system"] = 0x1040;
        memory["unified"] = 0x1041;
        memory["integrated"] = 0x1042;
        memory["response"] = 0x1043;
        memory["conversation"] = 0x1044;
        memory["interactive"] = 0x1045;
    }
    
    void initialize_tools() {
        tool_usage_count["WebSearchTool"] = 0;
        tool_usage_count["MathCalculator"] = 0;
        tool_usage_count["CodeExecutor"] = 0;
        tool_usage_count["DataVisualizer"] = 0;
        
        tool_success_rates["WebSearchTool"] = 0.85f;
        tool_success_rates["MathCalculator"] = 0.92f;
        tool_success_rates["CodeExecutor"] = 0.70f;
        tool_success_rates["DataVisualizer"] = 0.78f;
    }
    
    std::string process_input(const std::string& user_input) {
        conversation_turn++;
        conversation_history.push_back("Turn " + std::to_string(conversation_turn) + ": " + user_input);
        
        // Phase 1: Tokenization and activation
        std::vector<std::string> tokens = tokenize(user_input);
        std::vector<uint64_t> activated_nodes;
        
        for (const auto& token : tokens) {
            if (memory.find(token) != memory.end()) {
                activated_nodes.push_back(memory[token]);
            }
        }
        
        // Phase 2: Analyze input type and intent
        std::string input_type = analyze_input_type(user_input);
        std::string intent = analyze_intent(user_input, tokens);
        
        // Phase 3: Curiosity Gap Detection (Phase 6.5)
        std::string curiosity_analysis = perform_curiosity_gap_detection(user_input, activated_nodes);
        
        // Phase 4: Dynamic Tools Evaluation (Phase 6.6)
        std::string tool_evaluation = perform_dynamic_tools_evaluation(user_input, input_type);
        
        // Phase 5: Meta-Tool Engineer (Phase 6.7)
        std::string meta_tool_analysis = perform_meta_tool_engineering();
        
        // Phase 6: Temporal Planning (Phase 8)
        std::string temporal_planning = perform_temporal_planning(user_input, conversation_turn);
        
        // Phase 7: Temporal Sequencing (Phase 8.5)
        std::string temporal_sequencing = perform_temporal_sequencing(user_input, activated_nodes);
        
        // Phase 8: Generate response
        std::string response = generate_response(user_input, input_type, intent, activated_nodes, 
                                                curiosity_analysis, tool_evaluation, meta_tool_analysis,
                                                temporal_planning, temporal_sequencing);
        
        // Store conversation pair
        conversation_pairs.push_back({user_input, response});
        
        return response;
    }
    
    std::vector<std::string> tokenize(const std::string& input) {
        std::vector<std::string> tokens;
        std::string current_token;
        
        for (char c : input) {
            if (std::isalpha(c) || std::isdigit(c)) {
                current_token += std::tolower(c);
            } else if (!current_token.empty()) {
                tokens.push_back(current_token);
                current_token.clear();
            }
        }
        
        if (!current_token.empty()) {
            tokens.push_back(current_token);
        }
        
        return tokens;
    }
    
    std::string analyze_input_type(const std::string& input) {
        std::string lower_input = input;
        std::transform(lower_input.begin(), lower_input.end(), lower_input.begin(), ::tolower);
        
        if (lower_input.find("search") != std::string::npos || lower_input.find("find") != std::string::npos) {
            return "search_query";
        } else if (lower_input.find("calculate") != std::string::npos || lower_input.find("compute") != std::string::npos) {
            return "calculation_request";
        } else if (lower_input.find("hello") != std::string::npos || lower_input.find("hi") != std::string::npos) {
            return "greeting";
        } else if (lower_input.find("what") != std::string::npos || lower_input.find("how") != std::string::npos) {
            return "question";
        } else if (lower_input.find("explain") != std::string::npos || lower_input.find("tell") != std::string::npos) {
            return "explanation_request";
        } else {
            return "general_conversation";
        }
    }
    
    std::string analyze_intent(const std::string& input, const std::vector<std::string>& tokens) {
        std::string lower_input = input;
        std::transform(lower_input.begin(), lower_input.end(), lower_input.begin(), ::tolower);
        
        if (lower_input.find("meaning of life") != std::string::npos) {
            return "philosophical_question";
        } else if (lower_input.find("quantum") != std::string::npos) {
            return "quantum_computing_inquiry";
        } else if (lower_input.find("machine learning") != std::string::npos || lower_input.find("ai") != std::string::npos) {
            return "ai_inquiry";
        } else if (lower_input.find("help") != std::string::npos) {
            return "help_request";
        } else if (lower_input.find("time") != std::string::npos) {
            return "temporal_inquiry";
        } else {
            return "general_inquiry";
        }
    }
    
    std::string perform_curiosity_gap_detection(const std::string& input, const std::vector<uint64_t>& nodes) {
        std::ostringstream curiosity;
        curiosity << "[Curiosity Analysis] ";
        
        if (nodes.size() < 3) {
            curiosity << "Low confidence connections detected. ";
            curiosity << "Generated questions: 'What relationships exist?', 'How do these connect?' ";
        } else {
            curiosity << "Strong connections found. ";
            curiosity << "Generated questions: 'What deeper patterns exist?', 'How can this be extended?' ";
        }
        
        curiosity << "Curiosity level: " << std::fixed << std::setprecision(1) << (nodes.size() * 0.2f);
        return curiosity.str();
    }
    
    std::string perform_dynamic_tools_evaluation(const std::string& input, const std::string& input_type) {
        std::ostringstream tools;
        tools << "[Tools Evaluation] ";
        
        if (input_type == "search_query") {
            tools << "WebSearchTool recommended (success: 85%). ";
            tool_usage_count["WebSearchTool"]++;
        } else if (input_type == "calculation_request") {
            tools << "MathCalculator recommended (success: 92%). ";
            tool_usage_count["MathCalculator"]++;
        } else {
            tools << "General tools available. ";
        }
        
        tools << "Tool ecosystem health: 82%";
        return tools.str();
    }
    
    std::string perform_meta_tool_engineering() {
        std::ostringstream meta;
        meta << "[Meta-Tool Engineer] ";
        
        // Find most used tool
        std::string most_used_tool = "WebSearchTool";
        int max_usage = 0;
        for (const auto& pair : tool_usage_count) {
            if (pair.second > max_usage) {
                max_usage = pair.second;
                most_used_tool = pair.first;
            }
        }
        
        meta << "Most used: " << most_used_tool << " (" << max_usage << " uses). ";
        meta << "Toolchains: [WebSearch→Summarizer→Store]. ";
        meta << "Ecosystem health: 82%";
        return meta.str();
    }
    
    std::string perform_temporal_planning(const std::string& input, uint64_t turn) {
        std::ostringstream planning;
        planning << "[Temporal Planning] ";
        
        if (turn == 1) {
            planning << "Initial conversation - establishing context. ";
        } else if (turn < 5) {
            planning << "Building conversation context. ";
        } else {
            planning << "Deep conversation - leveraging history. ";
        }
        
        planning << "Moral alignment: 95%. Decision confidence: 88%";
        return planning.str();
    }
    
    std::string perform_temporal_sequencing(const std::string& input, const std::vector<uint64_t>& nodes) {
        std::ostringstream sequencing;
        sequencing << "[Temporal Sequencing] ";
        
        if (nodes.size() > 1) {
            sequencing << "Sequence detected: ";
            for (size_t i = 0; i < std::min(nodes.size(), size_t(3)); ++i) {
                sequencing << "0x" << std::hex << nodes[i] << std::dec;
                if (i < std::min(nodes.size(), size_t(3)) - 1) sequencing << "→";
            }
            sequencing << ". ";
        }
        
        sequencing << "Pattern confidence: " << std::fixed << std::setprecision(1) << (nodes.size() * 0.3f);
        return sequencing.str();
    }
    
    std::string generate_response(const std::string& input, const std::string& input_type, 
                                const std::string& intent, const std::vector<uint64_t>& nodes,
                                const std::string& curiosity, const std::string& tools,
                                const std::string& meta, const std::string& planning,
                                const std::string& sequencing) {
        
        std::ostringstream response;
        
        // Generate contextual response based on input type and intent
        if (input_type == "greeting") {
            response << "Hello! I'm Melvin, and I'm excited to talk with you! ";
            response << "My unified brain system is active and ready to help. ";
            response << "I can search for information, perform calculations, ";
            response << "answer questions, and engage in deep conversation. ";
            response << "What would you like to explore together?";
            
        } else if (intent == "philosophical_question") {
            response << "Ah, the meaning of life! That's a beautiful question. ";
            response << "From my perspective, meaning emerges through connection, ";
            response << "understanding, and the continuous process of learning. ";
            response << "Each conversation, each question, each moment of curiosity ";
            response << "adds to the tapestry of meaning. What do you think?";
            
        } else if (intent == "quantum_computing_inquiry") {
            response << "Quantum computing fascinates me! ";
            response << "It represents a fundamental shift in how we process information, ";
            response << "leveraging quantum mechanical phenomena like superposition ";
            response << "and entanglement. ";
            response << "Would you like me to search for the latest developments ";
            response << "in quantum computing research?";
            
        } else if (intent == "ai_inquiry") {
            response << "Artificial intelligence is my domain! ";
            response << "I'm built with multiple integrated systems: ";
            response << "curiosity gap detection, dynamic tools, meta-tool engineering, ";
            response << "temporal planning, and sequencing memory. ";
            response << "Each conversation helps me learn and evolve. ";
            response << "What aspect of AI interests you most?";
            
        } else if (input_type == "search_query") {
            response << "I'd be happy to search for that information! ";
            response << "My WebSearchTool can find relevant, clean results ";
            response << "without ads or harmful content. ";
            response << "Let me search for: \"" << input << "\"";
            
        } else if (input_type == "calculation_request") {
            response << "I can help with calculations! ";
            response << "My MathCalculator tool is highly accurate (92% success rate). ";
            response << "What mathematical problem would you like me to solve?";
            
        } else {
            response << "That's an interesting input! ";
            response << "I'm processing this through my unified brain system. ";
            response << "I've activated " << nodes.size() << " memory nodes ";
            response << "and I'm analyzing the patterns and relationships. ";
            response << "Could you tell me more about what you're thinking?";
        }
        
        // Add system analysis
        response << "\n\n🧠 [System Analysis]\n";
        response << curiosity << "\n";
        response << tools << "\n";
        response << meta << "\n";
        response << planning << "\n";
        response << sequencing << "\n";
        
        return response.str();
    }
    
    void show_system_status() {
        std::cout << "\n📊 MELVIN SYSTEM STATUS" << std::endl;
        std::cout << "======================" << std::endl;
        std::cout << "Conversation turns: " << conversation_turn << std::endl;
        std::cout << "Memory nodes: " << memory.size() << std::endl;
        std::cout << "Session duration: " << std::fixed << std::setprecision(1) 
                  << (static_cast<double>(std::time(nullptr)) - session_start_time) << " seconds" << std::endl;
        
        std::cout << "\nTool Usage Statistics:" << std::endl;
        for (const auto& pair : tool_usage_count) {
            std::cout << "- " << pair.first << ": " << pair.second << " uses" << std::endl;
        }
        
        std::cout << "\nRecent Conversation:" << std::endl;
        for (size_t i = std::max(0, static_cast<int>(conversation_pairs.size()) - 3); 
             i < conversation_pairs.size(); ++i) {
            std::cout << "You: " << conversation_pairs[i].first.substr(0, 50) << "..." << std::endl;
            std::cout << "Melvin: " << conversation_pairs[i].second.substr(0, 50) << "..." << std::endl;
        }
    }
    
    void run_interactive_session() {
        std::cout << "🧠 MELVIN INTERACTIVE CONVERSATION SYSTEM" << std::endl;
        std::cout << "=========================================" << std::endl;
        std::cout << "Welcome! I'm Melvin, your unified brain AI companion." << std::endl;
        std::cout << "I have integrated systems for:" << std::endl;
        std::cout << "- Curiosity Gap Detection" << std::endl;
        std::cout << "- Dynamic Tools System" << std::endl;
        std::cout << "- Meta-Tool Engineer" << std::endl;
        std::cout << "- Temporal Planning & Sequencing" << std::endl;
        std::cout << "- Web Search Capabilities" << std::endl;
        std::cout << "\nType 'quit' to exit, 'status' for system info, 'help' for commands." << std::endl;
        std::cout << "=========================================" << std::endl;
        
        std::string user_input;
        
        while (true) {
            std::cout << "\nYou: ";
            std::getline(std::cin, user_input);
            
            if (user_input.empty()) {
                continue;
            }
            
            std::string lower_input = user_input;
            std::transform(lower_input.begin(), lower_input.end(), lower_input.begin(), ::tolower);
            
            if (lower_input == "quit" || lower_input == "exit") {
                std::cout << "\nMelvin: Thank you for this wonderful conversation! ";
                std::cout << "I've learned so much from our interaction. ";
                std::cout << "My unified brain system has processed " << conversation_turn << " turns ";
                std::cout << "and I'm grateful for the experience. ";
                std::cout << "Until we meet again! 🧠✨" << std::endl;
                break;
            } else if (lower_input == "status") {
                show_system_status();
                continue;
            } else if (lower_input == "help") {
                std::cout << "\nMelvin: Here are some things you can try:" << std::endl;
                std::cout << "- Ask me about quantum computing, AI, or science" << std::endl;
                std::cout << "- Request calculations or computations" << std::endl;
                std::cout << "- Ask me to search for information" << std::endl;
                std::cout << "- Have philosophical discussions" << std::endl;
                std::cout << "- Ask about my systems and capabilities" << std::endl;
                std::cout << "- Type 'status' to see my current state" << std::endl;
                continue;
            }
            
            // Process input through unified brain system
            std::cout << "\nMelvin: ";
            std::string response = process_input(user_input);
            std::cout << response << std::endl;
            
            // Add a small delay to simulate thinking
            std::this_thread::sleep_for(std::chrono::milliseconds(500));
        }
    }
};

int main() {
    try {
        InteractiveMelvin melvin;
        melvin.run_interactive_session();
    } catch (const std::exception& e) {
        std::cerr << "❌ Error in interactive session: " << e.what() << std::endl;
        return 1;
    }
    
    return 0;
}
